Fix build issues

greg0ire opened the pull request:
Decorated text used to be wrapped too early in SymfonyStyle->block()
See https://github.com/symfony/symfony/pull/40348

greg0ire committed:
Accommodate 2 behaviors of symfony/console in test

Decorated text used to be wrapped too early in SymfonyStyle->block()
See https://github.com/symfony/symfony/pull/40348
The fix was not contributed to version 3, which means we have to rewrite
the test so that it passes for both the correct and the buggy version.

greg0ire commented:
Symfony v3 is in security-fixes mode only, that's why the bugfix wasn't contributed to v3

greg0ire commented:
The remaining failures looks like it might happen every month, or every year, but will not happen tomorrow.

greg0ire commented:
See @lcobucci fightying it 3 years ago in 46c0861f4542324061fa650545e389dd1a21d859 , in a commit hilariously named "Fix date calculation in tests (again)" :rofl:

greg0ire commented:
```
>>> (new DateTimeImmutable())->modify('-1 month')
=> DateTimeImmutable {#213
     +"date": "2021-03-02 09:21:48.312300",
     +"timezone_type": 3,
     +"timezone": "Europe/Paris",
   }
```

but-why.gif

matks commented:
> The remaining failures looks like it might happen every month, or every year, but will not happen tomorrow.

Hello, it might be naive and maybe I am the 47th person talking about it, but have you considered wrapping time behind an interface so you control it, and use PHP DateTime as an implementation of this interface?

https://www.matheus.ro/2017/09/25/unit-test-time-dependency/

beberlei commented:
https://twitter.com/stauffermatt/status/1376699768328630272?s=21

greg0ire commented:
@matks that would work for the PHP part, the trouble is, the other thing we are getting a date from is the RDBMS under test. Consider sending a PR in case you know how to fix that.

matks commented:
> @matks that would work for the PHP part, the trouble is, the other thing we are getting a date from is the RDBMS under test. Consider sending a PR in case you know how to fix that.

😅  I'm afraid that the RDBMS is instead a much bigger issue to handle.

I will try to look at it but I doubt my experience will be enough.

greg0ire commented:
I guess all we have to do is see how `subMonthsNoOverflow()` in @beberlei 's referenced tweet is implemented.

greg0ire committed:
Adapt test logic to PHP and SQLite

There seems to be at least 2 camps in the software world when it comes
to the question "What's today minus one month", today being at the end
of march.

While PHP and SQLite agree that that would be the 2nd of March, other
RDBMS than SQLite and humans will tell you that it's the last day of
February.

This patch ensures that we check one logic for SQLite, and the other
logic for other platforms.

matks commented:
> I pushed what I think is a solution but maybe it's not a great solution

It seems to be a pragmatic solution.

Moreover, if tomorrow SQLite behavior changes or MariaDB behavior changes, you might need to enforce even more not-portability. You are constrained by the RDBMS, so I dont think you need to give much thought to this topic. The test path is messy, but it's not because it's a bad test. It's because it represents a messy real-life situation 😅  .

lcobucci commented:
@greg0ire alternatively you can use a fixed date/time instead of relying on the current db date. That should make things more stable.

greg0ire commented:
@lcobucci I love that solution, why didn't I think of that? That way, if we really feel we want to enforce a specific behavior for `30-02-YYYY`, we can do so in a separate test, and we will no longer have a test that depends on the current time.

greg0ire commented:
Well now Postgresql is unhappy with this :(

greg0ire commented:
Reverting to the original solution, but do advise if you know how to use a fixed date in a way that will work with all platforms so as a string with most, but with `::timestamp` appended to it for PostgreSQL.

ostrolucky commented:
Not a great solution. Abstraction layer is supposed to attempt to abstract away these database differences. This test demonstrates it's not doing its job. Instead of hiding this fact in test case, solution could be to make sure all databases return same result given same parameters. This can be done perhaps on DateSubFunction level or in DBAL.

But I realize I ask for too much here, so I have different proposition how to avoid this problem every march. Stop testing CURRENT_TIMESTAMP() function, that's not the goal of this testscase and is already covered in multitude of the other tests. In this test case, always use value for first day of current month as a basis for DATE_SUB. Or if you don't like that much, random day in range of 1-28 of current month. Not the current date.

greg0ire commented:
@ostrolucky that's kinda what @lcobucci proposed, and I tried implementing it, but failed because of https://github.com/doctrine/orm/pull/8573#issuecomment-811094531 . Do you know of a workaround for that?

lcobucci commented:
@greg0ire I'd say that it reveals an issue on DBAL (the method that returns the date_* function should cast the string into a timestamp when that's necessary: https://www.db-fiddle.com/f/4jyoMCicNSZpjMt4jFYoz5/0)

ostrolucky commented:
Yeah. Workaround might be to have entry with hardcoded date in CompanyManager, then you can use something like `DATE_SUB(m.date, ...`. I can't decide if it's much better though.

greg0ire commented:
Ok well, let's go with the ugly solution that works then.
